src/test/java/software/amazon/awssdk/crt/test/HttpConnectionTest.java

```diff
@@ -18,33 +18,38 @@
 import org.junit.Assert;
 import org.junit.Test;
 import software.amazon.awssdk.crt.CrtResource;
-import software.amazon.awssdk.crt.CrtRuntimeException;
 import software.amazon.awssdk.crt.http.HttpConnection;
 import software.amazon.awssdk.crt.io.ClientBootstrap;
 import software.amazon.awssdk.crt.io.SocketOptions;
 import software.amazon.awssdk.crt.io.TlsContext;
 
 import java.net.URI;
+import java.util.concurrent.TimeUnit;
 
 public class HttpConnectionTest {
 
-    protected void testConnection(URI uri, boolean expectConnected, String exceptionMsg) throws CrtRuntimeException {
+    protected void testConnection(URI uri, boolean expectConnected, String exceptionMsg) throws Exception {
         boolean actuallyConnected = false;
         boolean exceptionThrown = false;
 
         ClientBootstrap bootstrap = new ClientBootstrap(1);
         SocketOptions sockOpts = new SocketOptions();
         TlsContext tlsContext =  new TlsContext();
 
+        HttpConnection conn = null;
         try {
-            HttpConnection conn = HttpConnection.createConnection(uri, bootstrap, sockOpts, tlsContext).get();
+            conn = HttpConnection.createConnection(uri, bootstrap, sockOpts, tlsContext).get();
             actuallyConnected = true;
-            conn.shutdown().get();
-            conn.close();
+
         } catch (Exception e) {
             exceptionThrown = true;
             Assert.assertTrue(e.getMessage(), e.getMessage().contains(exceptionMsg));
         } finally {
+            if (conn != null) {
+                conn.shutdown().get();
```

Comment:
Why not just have ```close()``` do this? This won't protect users in the wild.

Reply:
If `close()` automatically waited for `shutdown()` to complete then [this line](https://github.com/awslabs/aws-crt-java/pull/57/files#diff-dbee0f3e9747085aa04f4f613340a732R215) would block the EventLoop.

Reply:
I'm not sure that's true. If the future has already completed, you'll have no wait at all, and it'll throw, right? Am I missing something?

Reply:
> If the future has already completed

The shutdown future is [guaranteed to not be complete](https://github.com/awslabs/aws-c-http/blob/92c780a4f5bb20c54b91efa9bbe7b3988aecd6fa/include/aws/http/connection.h#L100-L101) when the connection fails to connect.

I'm trying to avoid calling and waiting for the `on_shutdown()` callback to complete while inside the `on_conn_complete()` callback. Users don't need to wait for shutdown before releasing the connection in that case, but should wait for shutdown in other cases.

Reply:
Addressed both of our issues. `close()` now automatically waits for `shutdown()` so users don't accidentally shoot themselves in the foot, but I also added logic skip shutdown [if we never successfully connected in the first place](https://github.com/awslabs/aws-crt-java/commit/9ee099d83b756a65572a0c1e722003ce48326d0d#diff-dbee0f3e9747085aa04f4f613340a732R181).